### PR TITLE
System uptime tracking

### PR DESCRIPTION
**Pull request by tbazant**
### PR creator: Description

New feature that enables clients' uptime tracking

### CODE REVIEW

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+          which is triggered by the
+          <systemitem class="daemon">suseconnect-keepalive</systemitem> timer
+          every 24 hours. Verify that the timer is running.
+        </para>
+<screen>&prompt.user;systemctl status suseconnect-keepalive.timer</screen>
+      </step>
+    </procedure>
+  </sect1>
+  <sect1 xml:id="uptime-data-retrieve">
+    <title>Retrieve the uptime tracker data</title>
```
Comment (dariavladykina) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:127]:
```suggestion
    <title>Retrieving the uptime tracker data</title>
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+            understanding.
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term>Online At Hours</term>
+        <listitem>
+          <para>
+            Each bit represents the hour of the day. The very left bit
+            corresponds the zeroth hour of the day. <quote>1</quote> indicates
```
Comment (dariavladykina) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:255]:
```suggestion
            Each bit represents the hour of the day. The leftmost position 
            represents the zeroth hour of the day. <quote>1</quote> indicates
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    </variablelist>
+  </sect1>
+  <sect1 xml:id="uptime-data-practical">
+    <title>Practical tasks</title>
+
+    <qandaset>
+      <qandaentry>
+        <question>
+          <para>
+            Was the host up between 9-10am on 2024-02-27?
```
Comment (dariavladykina) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:278]:
```suggestion
            Was the host up between 9–10 a.m. on 2024-02-27?
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+          <para>
+            Retrieve the uptime data as described in
+            <xref
+            linkend="uptime-data-retrieve"/>. Open the data
+            file and search for the line that starts with
+            <literal>2024-02-27</literal>. Inspect that day's bit pattern, for
+            example, <literal>000011111111111111111111</literal> and count up to
+            the 10th digit (the bit starts with 0am) from left. The
+            corresponding bit is <literal>1</literal>, therefore the system was
+            up between 9-10am.
```
Comment (dariavladykina) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:291]:
```suggestion
            the 10th digit (the bit starts with 0 a.m.) from left. The
            corresponding bit is <literal>1</literal>, therefore the system was
            up between 9 and 10 a.m.
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+-u'<replaceable>USERNAME</replaceable>:<replaceable>PASSWORD</replaceable>' \
+'https://scc.suse.com/connect/organizations/system_uptimes?SystemId=16300735&amp;OnlineFrom=2024-02-01&amp;OnlineTo=2024-02-29' \
+-H 'accept: text/csv'
+System;Uuid;Online At Day;Online At Hours;Online For Hours
+16300735;78740b8c-7d23-42f8-b5fe-4776c47b9cb9;2024-04-24;000000000000000000000001;1
+16300735;78740b8c-7d23-42f8-b5fe-4776c47b9cb9;2024-04-25;110000000000000000000000;2
+</screen>
+          <para>
+            The sum of the very right column is 3, there fore the host was up
+            for 3 days in February 2024.
```
Comment (dariavladykina) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:335]:
```suggestion
            The sum in the rightmost column is 3, indicating the host was up
            for 3 days in February 2024. 
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+      </qandaentry>
+      <qandaentry>
+        <question>
+          <para>
+            How many days was the host up in February 2024?
+          </para>
+        </question>
+        <answer>
+          <para>
+            To find out statistics about the host in a specific month, adjust
```
Comment (dariavladykina) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:318]:
```suggestion
            To retrieve statistics about the host in a specific month, adjust
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+15892209;9e818a04-9e82-4584-8f41-b0f39a045526;2024-03-12;111110000000000111111111;14
+15892209;9e818a04-9e82-4584-8f41-b0f39a045526;2024-03-13;111111000000000111111111;15
+15892209;9e818a04-9e82-4584-8f41-b0f39a045526;2024-03-14;111000000000000000000000;3
+15892210;9e818a04-9e82-4584-8f41-b0f39a045526;2024-02-27;000011111111111111111111;20
+15892210;9e818a04-9e82-4584-8f41-b0f39a045526;2024-02-28;000011111111111111111111;20
+15892210;9e818a04-9e82-4584-8f41-b0f39a045526;2024-02-29;000011111111111111111111;20
+15892210;9e818a04-9e82-4584-8f41-b0f39a045526;2024-03-01;011111000000000011111111;13
+</screen>
+          <para>
+            As you can see, the list inclides system IDs from different hosts.
```
Comment (dariavladykina) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:369]:
```suggestion
            As you can see, the list includes system IDs from different hosts.
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+      <dm:editurl>https://github.com/SUSE/doc-unversioned/edit/main/system-uptime-tracking/xml/</dm:editurl>
+      <dm:translation>yes</dm:translation>
+    </dm:docmanager>
+    <abstract>
+      <para>
+        The system uptime tracking feature monitors when a client system is up.
+        It works on an hourly basis&mdash;the system is considered up if it was
+        running at any part of the given hour. The information is stored in
+        &scc; (SCC) and Repository Mirroring Tool (RMT), and serves as a source
+        for analytic purposes.
```
Comment (lagartoflojo) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:39]:
```suggestion
        for analytics purposes.
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+        <para>
+          The version of the <package>suseconnect-ng</package> package has to be
+          1.10 or newer. Verify it by running <command>suseconnect
+          --version</command>.
+        </para>
+      </listitem>
+      <listitem>
+        <para>
+          If the client system is behind RMT, the version of the
+          <package>rmt-server</package> package has to be 2.17 or newer. Verify
```
Comment (lagartoflojo) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:57]: Maybe it's a bit ambiguous that the rmt-server package has to be 2.17 or newer _on the RMT server itself_, and not on the client system. (The client system does not need  to have the rmt-server package installed at all)

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    <para>
+      You can retrieve the data that the uptime tracker collects from SCC and
+      export them to the CSV format via its REST API.
+    </para>
+
+    <procedure>
+      <step>
+        <para>
+          Obtain your organization mirroring credentials from the
+          <guimenu>Proxies</guimenu> tab in the SCC Web at
```
Comment (lagartoflojo) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:138]:
```suggestion
          <guimenu>Proxies</guimenu> tab in the SCC website at
```

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+          <step>
+            <para>
+              To retrieve the data using the <command>wget</command> command,
+              run the following:
+            </para>
+<screen>
+&prompt.user;wget --no-check-certificate \
+--user=<replaceable>USERNAME</replaceable> --password=<replaceable>PASSWORD</replaceable> \
+--header='accept: text/csv' \
+'https://scc.suse.com/connect/organizations/system_uptimes'
```
Comment (lagartoflojo) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:189]: the curl command includes params (systemid, online from, online to), but the wget command doesn't. is that intentional?

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+      configuration.
+    </para>
+
+    <procedure>
+      <title>Enabling the &systemd; timer</title>
+      <step>
+        <para>
+          Install the <package>suse-uptime-tracker</package> package.
+        </para>
+<screen>&prompt.sudo;zypper install suse-uptime-tracker</screen>
```
Comment (jeremy-moffitt) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:78]: we may need to specify where to download this from until MSC-788 is complete, as the suse-uptime-tracker package is not yet part of the regular SLE channels

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+        <para>
+          Edit <filename>/etc/SUSEConnect</filename> and enable the uptime
+          tracking.
+        </para>
+<screen>enable_system_uptime_tracking: true</screen>
+      </step>
+      <step>
+        <para>
+          After enabling the uptime tracking, the timer checks if the system is
+          running every 15 minutes and stores the findings in the log files
```
Comment (jeremy-moffitt) [inline system-uptime-tracking/xml/art_system-uptime-tracking.xml:114]: its a singular log file, I would s/files/file/ in this line and the one below, and s/These/this/ in the line below

**Review by jeremy-moffitt — COMMENTED** [review]
one minor nit, and a more significant question around the suse-uptime-tracker package. while it will eventually be included in the same repos that suseconnect-ng is in, right now thats pending legal review and the maintenance team, so beta customers will need to install the package from its location in OBS at https://build.opensuse.org/package/show/systemsmanagement:SCC/uptime-tracker

**Review by jeremy-moffitt — APPROVED** [review]
Looks good to me, thanks for addressing my feedback. I'll ping a couple of the engineers for additional reviews as well.

File: system-uptime-tracking/xml/art_system-uptime-tracking.xml (hunk 11 above)
Comment (yarunachalam) [reply system-uptime-tracking/xml/art_system-uptime-tracking.xml:199]: it is not intentional. Updated the [https://confluence.suse.com/display/CloudSolutions/White+paper+documentation](url) section for query parameter "Here is the example of wget command retrieves system uptime data for system 16300735"

Comment (yarunachalam) [reply system-uptime-tracking/xml/art_system-uptime-tracking.xml:199]: Updated the document to match curl query parameter date and system id
`wget --user=<username> --password=<Password>'[http://192.168.196.107:3000/connect/organizations/system_uptimes?SystemId=16300735&OnlineFrom=2024-04-24&OnlineTo=2024-05-01'](http://192.168.196.107:3000/connect/organizations/system_uptimes?SystemId=16300735&OnlineFrom=2024-04-24&OnlineTo=2024-05-01%27)`

**Pull request merged**